[stdlib] Adding new indices model tests

austinzheng opened the pull request:
#### What's in this pull request?

More tests for the new Swift 3 indexing model. Tests pass on OS X.

@gribozavr 

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

Changes:
- Added tests for 'index(next:)'
- Added tests for all 'formIndex' API variants
- Modified existing tests to better test bidirectional collections

moiseev commented on stdlib/private/StdlibCollectionUnittest/CheckCollectionType.swift.gyb:
```diff
@@ -349,15 +349,26 @@ public struct IndexStepsFromTest {
+  DistanceFromToTest(start: 19, end: 0)
 ]
 
 public let indexStepsFromTests = [
```
`indexOffsetByTests` maybe? ;-) Since the method has been renamed.

moiseev commented on stdlib/private/StdlibCollectionUnittest/CheckCollectionType.swift.gyb:
```diff
@@ -1549,24 +1573,61 @@ extension TestSuite {
       return makeCollection(r.map { wrapValue(OpaqueValue($0)) })
     }
+
+    self.test("\(testNamePrefix).index(after:)/semantics") {
```
This whole section looks perfectly GYBbable. There are only two things that change between the mutating and non-mutating variants: test name and the line that produces the next index.

austinzheng replied:
GYB is a good idea.

moiseev commented on stdlib/private/StdlibCollectionUnittest/CheckCollectionType.swift.gyb:
```diff
@@ -1580,54 +1641,149 @@ extension TestSuite {
+        if let expectedOffset = test.expectedOffset {
+          expectEqual(c.nthIndex(expectedOffset), new,
             stackTrace: SourceLocStack().with(test.loc))
+          expectEqual(true, exact, stackTrace: SourceLocStack().with(test.loc))
```
FYI: there is an `expectTrue` function.

moiseev commented on stdlib/private/StdlibCollectionUnittest/CheckCollectionType.swift.gyb:
```diff
@@ -1580,54 +1641,149 @@ extension TestSuite {
+        } else {
+          // Clamped to the limit
+          expectEqual(limit, new, stackTrace: SourceLocStack().with(test.loc))
+          expectEqual(false, exact, stackTrace: SourceLocStack().with(test.loc))
```
..and `expectFalse` as well.

moiseev commented on validation-test/stdlib/Index.swift.gyb:
```diff
@@ -64,7 +64,11 @@ Index.test(
   ) {
-    let c = ${Collection}(Array(0..<10))
+    let max = 10
+    guard test.expectedOffset! < max else {
```
A minor nitpick: there's already a `filter` call on line 65, it only feels natural to include this condition in there. What do you think, @austinzheng?

austinzheng replied:
Agree, that sounds like a good idea. I'll make that change.

austinzheng commented:
@moiseev 

I made all the changes you requested, rebased, and squashed. All the tests pass on my OS X machine. Let me know if you have any questions.

moiseev commented on stdlib/private/StdlibCollectionUnittest/CheckCollectionType.swift.gyb:
```diff
@@ -1549,24 +1573,53 @@ extension TestSuite {
       return makeCollection(r.map { wrapValue(OpaqueValue($0)) })
     }
+
+%   for i in ['index', 'formIndex']:
```
In general, I would prefer a more descriptive name. But I guess it's alright for a scope of 20 lines.

moiseev commented on validation-test/stdlib/Index.swift.gyb:
```diff
@@ -61,10 +61,11 @@ Index.test(
-    {$0.limit == nil && $0.distance >= 0}
+  let max = 10
+  for test in indexOffsetByTests.filter(
+    { $0.limit == nil && $0.distance >= 0 && $0.expectedOffset! < max }
```
What if `expectedOffset` is `nil`? It is probably not in any of the cases that exist now, but just to make it future-proof, can you please check for `nil`ness as well?
```Swift
$0.expectedOffset.map {$0 < max} ?? false
```

moiseev commented:
Other than the `nil`ness comment, the change LGTM. Thanks @austinzheng!

austinzheng commented:
I updated the diff based on your comments. Let me know if you want any additional work, or if you want me to squash the commits. @moiseev

moiseev commented on validation-test/stdlib/Index.swift.gyb:
```diff
@@ -60,7 +60,7 @@ Index.test(
   let max = 10
   for test in indexOffsetByTests.filter(
-    { $0.limit == nil && $0.distance >= 0 && $0.expectedOffset! < max }
+    { $0.limit == nil && $0.distance >= 0 && ($0.expectedOffset ?? max) < max }
```
:+1: Thanks!

moiseev commented:
@swift-ci Please test

moiseev commented:
Thanks @austinzheng, let's run the CI and if it's all good, I'll let github squash the commits for us.
